sympy/stats/stochastic_process_types.py

```diff
@@ -727,6 +731,34 @@ def limiting_distribution(self):
         """
         return self.fixed_row_vector()
 
+    def sample(self, max_time):
+        """
+        Parameters
+        ==========
+
+        max_time: Positive Integer
+            Maximum Time upto which iteration is to be done. By default is None
+
+        Returns
+        =======
+
+        sample: iterator object
+            iterator object containing the sample
+
+        """
+        if not isinstance(self.transition_probabilities, (Matrix, ImmutableMatrix)):
+            raise ValueError("Transition Matrix must be provided for sampling")
+        Tlist = self.transition_probabilities.tolist()
+        samps = [random.choice(list(self.state_space))]
+        yield samps[0]
+        time = 1
+        while time < max_time:
+            dist = Tlist[samps[time - 1]]
+            density = {list(self.state_space)[i]: dist[i] for i in range(len(dist))}
```

Comment:
Can you move this outside while loop?

Reply:
I see that, density for all the states can be computed before starting the sampling loop.